File: session5/print_graph.m
% Generate a graphviz script that plots a graph.
% Input: Adjacency matrix, "node states" (could be partition identity, for example), and a filename to print to.
% The node state vector shall contain one element for each node and this element can be 0, 1, 2 or 3.

% adjacency_matrix[NxN matrix]: an adjacency matrix describing the network
% node_states[Nx1 vector]: a vector with elements in {0,1,2,3}  
% filename[string]: name of the file to print to, should have extension .gv for 'plotallgraphs.sh' to work without modification

function print_graph(adjacency_matrix,node_states,filename)

nr_nodes=size(adjacency_matrix,1);
labels ={'A','B','C','D'};
colors={'red','blue','black','green'};

file1=fopen(filename,'w');
fprintf(file1,'graph G {\n');
%fprintf(file1,['label=\"' graph_label '";']);
%fprintf(file1,'graph  [overlap=false, orientation=portrait,splines=true,remincross=true];\n');
%fprintf(file1,'node [color=grey, style=filled];\n');
%fprintf(file1,'node [fontname=\"Verdana\", size=\"30,30\"];\n');

%'Define' all nodes
for current_node=1:nr_nodes
    current_node_label = ['"' num2str(current_node) '\n' labels{node_states(current_node)+1}  '"'  ' [color = ' colors{node_states(current_node)+1} ']'];
    fprintf(file1,'%s ; \n',current_node_label);
end

% Connect the nodes
for to_node=1:nr_nodes
    for from_node=to_node+1:nr_nodes
        if adjacency_matrix(to_node,from_node)
            label1 = ['"' num2str(to_node) '\n' labels{node_states(to_node)+1} '"'];
            label2 = ['"' num2str(from_node) '\n' labels{node_states(from_node)+1} '"'];
            fprintf(file1,'%s -- %s ; \n',label1,label2);
        end
    end
end
fprintf(file1,'}');
fclose(file1);

return

File: session5/print_graph.py
# Author: Sylvain Lapeyrade (sylla801)
# File: TSKS11-Lab5: Function print Graph


# Function translated from Matlab to Python
def print_graph(adjacency_matrix, node_states, filename) -> None:
    """
    Generate a graphviz script that plots a graph.
    Input: Adjacency matrix, "node states" (could be partition identity,
     for example), and a filename to print to.
    The node state vector shall contain one element for each node
     and this element can be 0, 1, 2 or 3.


    :param adjacency_matrix: A `numpy.ndarray` or a `scipy.sparce matrix`
    :param node_states: A numpy.ndarray of shape N
    :param filename: A string which the resulting graphviz file should be
     called. For maximum compatibility make sure it ends on .gv otherwise
      the `plotallgraphs.sh` does not work without modification.
    """
    nr_nodes = adjacency_matrix.shape[0]
    labels = ['A', 'B', 'C', 'D']
    colors = ['red', 'blue', 'black', 'green']

    with open(filename, 'w') as file1:
        file1.write('graph G {\n')
        file1.write('graph [overlap=false, orientation=portrait,' +
                    'splines=true, remincross=true];\n')
        file1.write('node [color=grey, style=filled];\n')
        file1.write('node [fontname=\"Verdana\", size=\"30,30\"];\n')

        for current_node in range(0, nr_nodes):
            current_node_label = '"{}\\n{}" [color = {}] ;\n'.format(current_node + 1,
                                                                     labels[node_states[current_node]],
                                                                     colors[node_states[current_node]])
            file1.write(current_node_label)
            for friend in adjacency_matrix[:, current_node].nonzero()[0]:
                link = '"{}\\n{}" -- "{}\\n{}";\n'.format(current_node + 1,
                                                          labels[node_states[current_node]],
                                                          friend + 1,
                                                          labels[node_states[friend]])
                file1.write(link)

        file1.write('}')
